jvb/src/main/kotlin/org/jitsi/videobridge/xmpp/ClientConnection.kt

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright @ 2018 - present 8x8, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jitsi.videobridge.xmpp
+
+import org.jitsi.nlj.stats.DelayStats
+import org.jitsi.nlj.util.OrderedJsonObject
+import org.jitsi.utils.logging2.cdebug
+import org.jitsi.utils.logging2.createLogger
+import org.jitsi.videobridge.xmpp.config.XmppClientConnectionConfig
+import org.jitsi.xmpp.extensions.colibri.ColibriConferenceIQ
+import org.jitsi.xmpp.extensions.colibri.ShutdownIQ
+import org.jitsi.xmpp.extensions.health.HealthCheckIQ
+import org.jitsi.xmpp.mucclient.IQListener
+import org.jitsi.xmpp.mucclient.MucClientConfiguration
+import org.jitsi.xmpp.mucclient.MucClientManager
+import org.jitsi.xmpp.util.IQUtils
+import org.jivesoftware.smack.packet.ExtensionElement
+import org.jivesoftware.smack.packet.IQ
+import org.jivesoftware.smack.packet.XMPPError
+import org.jivesoftware.smackx.iqversion.packet.Version
+import org.json.simple.JSONObject
+import java.util.concurrent.atomic.AtomicBoolean
+
+/**
+ * The XMPP client connection for the videobridge
+ */
+class ClientConnection : IQListener {
+    private val logger = createLogger()
+
+    /**
+     * The [MucClientManager] which manages the XMPP user connections and the MUCs.
+     */
+    val mucClientManager = MucClientManager(FEATURES)
+
+    private val config = XmppClientConnectionConfig()
+
+    private val running = AtomicBoolean(false)
+
+    var eventHandler: EventHandler? = null
+
+    fun start() {
+        if (running.compareAndSet(false, true)) {
+            mucClientManager.apply {
+                registerIQ(HealthCheckIQ())
+                registerIQ(ColibriConferenceIQ())
+                registerIQ(Version())
+                registerIQ(ShutdownIQ.createForceShutdownIQ())
+                registerIQ(ShutdownIQ.createGracefulShutdownIQ())
+                setIQListener(this@ClientConnection)
+            }
+
+            config.clientConfigs.forEach { cfg -> mucClientManager.addMucClient(cfg) }
+        } else {
+            logger.info("Already started")
+        }
+    }
+
+    fun stop() {
+        if (running.compareAndSet(true, false)) {
+            mucClientManager.stop()
+        }
+    }
+
+    fun setPresenceExtension(extension: ExtensionElement) {
+        mucClientManager.setPresenceExtension(extension)
+    }
+
+    fun addMucClient(jsonObject: JSONObject): Boolean {
```

Comment:
Can you bring back the docs from the original? They describe the json format

Reply:
Other docs in this class are also useful

Reply:
Oops, definitely.